tests/queries/0_stateless/02156_storage_merge_prewhere.reference

```diff
@@ -0,0 +1,13 @@
+SELECT count()
+FROM t_02156_merge1
+PREWHERE k = 1
```

Comment:
Actually it should not move `k` AFAIU, since it is in the `PRIMARY KEY` (`ORDER BY`), but it does so because StorageMerge does not know about primary key of the MergeTree, and so MergeTreeWhereOptimizer works incorrectly.

Plus, right now this test [fails for wide parts](https://s3.amazonaws.com/clickhouse-test-reports/33392/1e9904555a482f3fb586a3174b5e42335db51678/stateless_tests__release__wide_parts_enabled__actions_.html)

@CurtizJ maybe StorageMerge should pass through primary key to the MergeTreeWhereOptimizer, and do this optimization only if the primary key matches?

Reply:
You are right. But it's true not for all primary key columns, but only for first column from primary key. And also the behaviour differs when there is `FINAL` in query.